### PR TITLE
feat: add query-session-end and improve session-end events on Windows

### PR DESCRIPTION
**Pull request by savely-krasovsky**
#### Description of Change

Current Electron version allows to handle only `WM_ENDSESSION`, but not `WM_QUERYENDSESSION`. This cause a lot of problems while Electron app is packaged into own MSI and is a menubar/tray app. For example currently we cannot close app using Restart Manager which is used by default since Windows Installer 4.0. We have to use [util:CloseApplication](https://wixtoolset.org/docs/schema/util/closeapplication/) or call `taskkil`.

Quote from [Restart Manager guidelines for Applications](https://learn.microsoft.com/en-us/windows/win32/rstmgr/guidelines-for-applications):

> Applications running on Windows Vista and Windows Server 2008 should adhere to these guidelines to ensure that the Restart Manager can shut down and restart applications if necessary to install updates.

> GUI applications should listen for the WM_QUERYENDSESSION message and return a value of TRUE if the application is prepared to shut down and restart. If no application returns a value of FALSE, the Restart Manager sends a [WM_ENDSESSION](https://learn.microsoft.com/en-us/windows/desktop/Shutdown/wm-endsession) message with the lParam parameter set to ENDSESSION_CLOSEAPP (0x1) and the wparam parameter set to TRUE. [...] If any GUI application responds to a WM_QUERYENDSESSION message by returning a value of FALSE, the shutdown is canceled.

So to properly close GUI app we need to handle `WM_QUERYENDSESSION` properly and return `TRUE`. But for now for any unknown WM message it will cause us to return FALSE (see calling code: https://github.com/chromium/chromium/blob/134896c2d6148c6bd2f7b096b0664b1e4013cb5c/ui/views/win/hwnd_message_handler.cc#L1069).

My implementation will return TRUE by default (in case of `WM_QUERYSESSION`), so it could probably fix longstanding issues like this: https://github.com/electron/electron/issues/15880. `ev.preventDefault()` can still return old behaviour.

Also both `query-session-end` and `session-end` now return a reasons why they receiving this event, it could be `shutdown`, `close-app`, `critical` and `logoff`.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Added `query-session-end` event and improved `session-end` events on Windows

### CODE REVIEW

File: docs/tutorial/multithreading.md [hunk 1]
```diff
@@ -42,7 +42,7 @@ safe.
 The only way to load a native module safely for now, is to make sure the app
 loads no native modules after the Web Workers get started.
 
-```js @ts-expect-error=[1]
```
Comment (itsananderson) [inline docs/tutorial/multithreading.md:45]: This change seems unrelated to the others. Is this required?

File: docs/api/base-window.md [hunk 2]
```diff
@@ -144,10 +144,27 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
...
 remove the reference to the window and avoid using it any more.
 
+#### Event: 'query-session-end' _Windows_
+
+Returns:
+
+* `event` Event
+* `reasons` string[]
+
+Emitted when window session is going to end due to force shutdown or machine
```
Comment (itsananderson) [inline docs/api/base-window.md:154]: This mentions "force shutdown". Does that mean `query-session-end` will fire again if the user elects to force a shutdown in the "These apps are preventing shutdown" screen? I would have intuitively expected that to trigger the `session-end` event instead.

File: docs/tutorial/multithreading.md (hunk 1 above)
Comment (savely-krasovsky) [reply docs/tutorial/multithreading.md:45]: I wasn't able to commit because of some pre-commit hook that was triggered on that. I can remove that change and ignore pre-commit hook error probably.

File: docs/api/base-window.md (hunk 2 above)
Comment (savely-krasovsky) [reply docs/api/base-window.md:154]: No, it won't I will remove "force" word.

**Review by codebytere — CHANGES_REQUESTED** [review]
Thanks for working on this!

**Comment by savely-krasovsky** [comment]
@codebytere thanks for the review! Everything sounds reasonable. I would commit fixes soon.

File: shell/browser/native_window_views_win.cc [hunk 3]
```diff
@@ -27,6 +27,24 @@ namespace electron {
 
 namespace {
 
+// Convert Win32 WM_QUERYENDSESSIONS to strings.
+const std::vector<std::string> EndSessionToStringVec(LPARAM end_session_id) {
+  std::vector<std::string> params;
+  if (end_session_id == 0) {
+    params.push_back("shutdown");
```
[the remark this answers is not in the document]
Comment (savely-krasovsky) [reply shell/browser/native_window_views_win.cc:34]: In my opinion good documentation is enough. By `1)` do you mean that `shutdown` reason should be named something like `shutdown-or-reboot`?

[the remark this answers is not in the document]
Comment (savely-krasovsky) [reply shell/browser/native_window_views_win.cc:34]: I improved the docs, should be clear enough now in my opinion.

File: docs/api/base-window.md [hunk 4]
```diff
@@ -144,10 +144,34 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
 Emitted when the window is closed. After you have received this event you should
 remove the reference to the window and avoid using it any more.
 
+#### Event: 'query-session-end' _Windows_
+
+Returns:
+
+* `event` Event
+* `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.
```
Comment (samuelmaddock) [inline docs/api/base-window.md:152]: Our documentation parser will use string literals with the "Can be ..." syntax.
```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

File: docs/api/base-window.md [hunk 5]
```diff
@@ -144,10 +144,34 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
...
+refer to the [MSDN documentation](https://learn.microsoft.com/en-us/windows/win32/shutdown/wm-queryendsession).
+
 #### Event: 'session-end' _Windows_
 
-Emitted when window session is going to end due to force shutdown or machine restart
-or session log off.
+Returns:
+
+* `event` Event
+* `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.
```
Comment (samuelmaddock) [inline docs/api/base-window.md:168]:
```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

File: docs/api/browser-window.md [hunk 6]
```diff
@@ -207,10 +207,34 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
 Emitted when the window is closed. After you have received this event you should
 remove the reference to the window and avoid using it any more.
 
+#### Event: 'query-session-end' _Windows_
+
+Returns:
+
+* `event` Event
+* `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.
```
Comment (samuelmaddock) [inline docs/api/browser-window.md:215]:
```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

File: docs/api/browser-window.md [hunk 7]
```diff
@@ -207,10 +207,34 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
...
+refer to the [MSDN documentation](https://learn.microsoft.com/en-us/windows/win32/shutdown/wm-queryendsession).
+
 #### Event: 'session-end' _Windows_
 
-Emitted when window session is going to end due to force shutdown or machine restart
-or session log off.
+Returns:
+
+* `event` Event
+* `reasons` string[] - Can contain a combination of those reasons 'shutdown', 'close-app', 'critical' and 'logoff'.
```
Comment (samuelmaddock) [inline docs/api/browser-window.md:231]:
```suggestion
* `reasons` string[] - List of reasons for shutdown. Can be 'shutdown', 'close-app', 'critical', or 'logoff'.
```

File: shell/browser/api/electron_api_base_window.cc [hunk 8]
```diff
@@ -177,8 +177,15 @@ void BaseWindow::OnWindowClosed() {
...
-  Emit("session-end");
+void BaseWindow::OnWindowQueryEndSession(const std::vector<std::string> reasons,
+                                         bool* prevent_default) {
+  if (Emit("query-session-end", reasons)) {
+    *prevent_default = true;
+  }
+}
+
+void BaseWindow::OnWindowEndSession(const std::vector<std::string> reasons) {
+  Emit("session-end", reasons);
```
Comment (samuelmaddock) [inline shell/browser/api/electron_api_base_window.cc:188]: Pass by reference. Please also update other interfaces where `reasons` is passed in.
```suggestion
void BaseWindow::OnWindowQueryEndSession(const std::vector<std::string>& reasons,
                                         bool* prevent_default) {
  if (Emit("query-session-end", reasons)) {
    *prevent_default = true;
  }
}

void BaseWindow::OnWindowEndSession(const std::vector<std::string>& reasons) {
  Emit("session-end", reasons);
```

**Review by samuelmaddock — CHANGES_REQUESTED** [review]

File: shell/browser/api/electron_api_base_window.cc [hunk 9]
```diff
@@ -177,8 +177,15 @@ void BaseWindow::OnWindowClosed() {
       FROM_HERE, GetDestroyClosure());
 }
 
-void BaseWindow::OnWindowEndSession() {
-  Emit("session-end");
+void BaseWindow::OnWindowQueryEndSession(const std::vector<std::string> reasons,
+                                         bool* prevent_default) {
+  if (Emit("query-session-end", reasons)) {
+    *prevent_default = true;
+  }
```
Comment (samuelmaddock) [inline shell/browser/api/electron_api_base_window.cc:184]: We have a newer pattern where we expose the values on the Event object directly. Can you change this event and the other to use this approach? Here's what it'd look like here: 
```suggestion
  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
  v8::HandleScope handle_scope(isolate);
  gin::Handle<gin_helper::internal::Event> event =
      gin_helper::internal::Event::New(isolate);
  v8::Local<v8::Object> event_object = event.ToV8().As<v8::Object>();
  gin::Dictionary dict(isolate, event_object);
  dict.Set("reasons", reasons);
  EmitWithoutEvent("query-session-end", event);
  if (event->GetDefaultPrevented()) {
    *prevent_default = true;
  }
```

To see how to document this, take a look at the changes to expose the 'audio-state-changed' event: https://github.com/electron/electron/commit/512e56baf75bce86aef11184af881cff67d8e6a2

**Review by samuelmaddock — CHANGES_REQUESTED** [review]

File: shell/browser/api/electron_api_base_window.cc (hunk 9 above)
Comment (savely-krasovsky) [reply shell/browser/api/electron_api_base_window.cc]: @samuelmaddock I tried to document it the same way, but unfortunately I am getting a lot of errors during `electron-api.json` generation:
```
../electron.d.ts(3898,5): error TS2416: Property 'on' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
../electron.d.ts(3900,5): error TS2416: Property 'off' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
../electron.d.ts(3902,5): error TS2416: Property 'once' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
[...]
```

I would suggest it's because all BaseWindow events should be ported in the same way.

Comment (savely-krasovsky) [reply shell/browser/api/electron_api_base_window.cc]: I forget escaping, sorry 🤦

Comment (savely-krasovsky) [reply shell/browser/api/electron_api_base_window.cc]: I found the root of problem: after adding event to both `base-window.md` and `browser-window.md` it will generate four types:
- BaseWindowQuerySessionEndEventParams
- BaseWindowSessionEndEventParams
- BrowserWindowQuerySessionEndEventParams
- BrowserWindowSessionEndEventParams

After manually removing two latter types from `electron.d.td` it started to work, but I don't know how to fix it without fixing typing generator.

Comment (samuelmaddock) [reply shell/browser/api/electron_api_base_window.cc]: This does unfortunately seem like something that needs to be fixed with our type generator. We want BrowserWindow docs on the website to show everything in BaseWindow so we've been duplicating the docs for now. A real fix might involve changing how our website is generated to include inherited methods.

For the purpose of this PR though, I think we could just have these types in BrowserWindow so that it's included in the BrowserWindow docs on the website as well. I'd be curious if other folks in @electron/wg-api have thoughts on how to address this.

Comment (itsananderson) [reply shell/browser/api/electron_api_base_window.cc]: What if you define a structure for those events, and reference the same structure(s) from both `BrowserWindow` and `BaseWindow`?

An example of such a structure that extends Event can be found here: https://github.com/electron/electron/blob/main/docs/api/structures/ipc-main-event.md

If you did that, then maybe TypeScript would consider the types compatible?

Comment (savely-krasovsky) [reply shell/browser/api/electron_api_base_window.cc]: @samuelmaddock there is another problem. I am not C++ expert, but `BaseWindow` C++ class inherits `gin_helper::TrackableObject` which seems like cannot be used with `gin_helper::EventEmitterMixin` which I need to use `EmitWithoutEvent`.

My guess is that probably `BaseWindow` class should be fully refactored to be compatible with `EventEmitterMixin`.

Comment (savely-krasovsky) [reply shell/browser/api/electron_api_base_window.cc]: I pushed a very naive fix, it compiles properly, but I am not sure this is the way.

File: shell/common/gin_helper/event_emitter.h [hunk 10]
```diff
@@ -65,6 +57,13 @@ class EventEmitter : public gin_helper::Wrappable<T> {
                           std::forward<Args>(args)...);
     return event->GetDefaultPrevented();
   }
+
+  // disable copy
+  EventEmitter(const EventEmitter&) = delete;
+  EventEmitter& operator=(const EventEmitter&) = delete;
+
+ protected:
+  EventEmitter() {}
```
[the remark this answers is not in the document]
Comment (savely-krasovsky) [reply shell/common/gin_helper/event_emitter.h]: @samuelmaddock I hope I understood you right, commited changes.

File: docs/api/base-window.md [hunk 11]
```diff
@@ -144,11 +144,6 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
 Emitted when the window is closed. After you have received this event you should
 remove the reference to the window and avoid using it any more.
 
-#### Event: 'session-end' _Windows_
```
Comment (samuelmaddock) [inline docs/api/base-window.md:147]: One thing I missed is that we probably don't want to outright remove this event if it can be avoided.

@savely-krasovsky were you able to give @itsananderson's suggestion a try?
https://github.com/electron/electron/pull/44598#discussion_r1842998014

**Comment by savely-krasovsky** [comment]
Any ideas what went wrong with Linux tests? I didn't change anything related.

File: docs/api/base-window.md [hunk 12]
```diff
@@ -144,10 +144,24 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
...
+
 #### Event: 'session-end' _Windows_
 
-Emitted when window session is going to end due to force shutdown or machine restart
-or session log off.
+Returns:
+
+* `event` [WindowSessionEndEvent][window-session-end-event]
+
+Emitted when system is in the process of shutting down after successfully handling 'query-session-end' event.
```
Comment (itsananderson) [inline docs/api/base-window.md:164]: Wording suggestion. "after successfully handling" doesn't seem quite right, because the user can still force a shutdown even if you `event.preventDefault()` the `query-session-end` event.

```suggestion
Emitted when a session is about to end due to a shutdown, machine restart, or user log-off. Once this event fires, there is no way to prevent the session from ending.
```

File: docs/api/browser-window.md [hunk 13]
```diff
@@ -207,10 +207,24 @@ _**Note**: There is a subtle difference between the behaviors of `window.onbefor
...
+
 #### Event: 'session-end' _Windows_
 
-Emitted when window session is going to end due to force shutdown or machine restart
-or session log off.
+Returns:
+
+* `event` [WindowSessionEndEvent][window-session-end-event]
+
+Emitted when system is in the process of shutting down after successfully handling 'query-session-end' event.
```
Comment (itsananderson) [inline docs/api/browser-window.md:227]: Wording suggestion

```suggestion
Emitted when a session is about to end due to a shutdown, machine restart, or user log-off. Once this event fires, there is no way to prevent the session from ending.
```

File: docs/api/base-window.md (hunk 11 above)
Comment (savely-krasovsky) [reply docs/api/base-window.md:147]: @samuelmaddock could you approve my PR if everything is ok now?

**Review by itsananderson — COMMENTED** [review]
Left a small suggestion for the `session-end` event, but overall API LGTM

File: docs/api/browser-window.md (hunk 13 above)
Comment (savely-krasovsky) [reply docs/api/browser-window.md]: Thanks for the suggesion, looks definitely better. Commited changes.

**Comment by savely-krasovsky** [comment]
@itsananderson @samuelmaddock @codebytere requesting a last review I hope! Fixed everything you highlighted.

**Review by itsananderson — APPROVED** [review]
Thanks for submitting this!

**Review by VerteDinde — APPROVED** [review]

**Review by codebytere — APPROVED** [review]
